
# Project: Investigate a Dataset - [TMDb Movie Data]

## Table of Contents
<ul>
<li><a href="#intro">Introduction</a></li>
<li><a href="#wrangling">Data Wrangling</a></li>
<li><a href="#eda">Exploratory Data Analysis</a></li>
<li><a href="#conclusions">Conclusions</a></li>
</ul>

<a id='intro'></a>
## Introduction

### Dataset Description 

<p>This data set contains information about 10,000 movies collected from The Movie Database (TMDb), including user ratings and revenue. (source: <a href = 'https://s3.amazonaws.com/video.udacity-data.com/topher/2018/July/5b57919a_data-set-options/data-set-options.pdf'>link</a>)</p>
<p>The table contains the following columns:</p>
<table>

<tr>
<th>Column</th>
<th>Significance</th>
</tr>
  
<tr>
<td>id</td>
<td>Unique ID for the row(associated movie)</td>
</tr>

<tr>
<td>imdb_id</td>
<td>Unique IMDB ID of the associated movie</td>
</tr>

<tr>
<td>popularity</td>
<td>Popularity score of the associated movie</td>
</tr>

<tr>
<td>budget</td>
<td>The budget of the associated movie</td>
</tr>

<tr>
<td>revenue</td>
<td>The revenue generated by the associated movie</td>
</tr>

<tr>
<td>original_title</td>
<td>The original title of the associated movie</td>
</tr>

<tr>
<td>cast</td>
<td>The associated movie cast names</td>
</tr>

<tr>
<td>homepage</td>
<td>The home page of the website of the associated movie </td>
</tr>

<tr>
<td>director</td>
<td>The name of the movie director of the associated movie</td>
</tr>

<tr>
<td>tagline</td>
<td>The associated movie tagline</td>
</tr>


<tr>
<td>keywords</td>
<td>Associated movie keywords</td>
</tr>

<tr>
<td>overview</td>
<td>Overview of the movie associated</td>
</tr>

<tr>
<td>runtime</td>
<td>Associated movie runtime in minutes</td>
</tr>

<tr>
<td>genres</td>
<td>The category associated movie</td>
</tr>

<tr>
<td>production_companies</td>
<td>Names of production companies of the associated movie</td>
</tr>

<tr>
<td>release_date</td>
<td>Release date of the associated movie</td>
</tr>

<tr>
<td>vote_count</td>
<td>Number of votes made by viewers on the associated movie</td>
</tr>

<tr>
<td>vote_average</td>
<td>Average vote qaulity fo the associated movie</td>
</tr>

<tr>
<td>release_year</td>
<td>Releae year of the associated movie</td>
</tr>

<tr>
<td>budget_adj</td>
<td>Budget of the associated movie in terms of 2010 dollars, accounting for inflation over time.</td>
</tr>

<tr>
<td>revenue_adj</td>
<td>Revenue of the associated movie in terms of 2010 dollars, accounting for inflation over time.</td>
</tr>

</table>


### Question(s) for Analysis
<ol>
    <li>Do voters give longer movies higher ratings?</li>
    <li>Which is the most popular genre?and by what percentage?</li>
    <li>Does total annual profit rise over time?</li>
    <li>Who are the top directors interms of profit made by moviesthey directed in the last decade?</li>
    <li>Which movie casts have been the most popular in the last decade?</li>
</ol>

In [1]:
#imports pandas, numpy, matplotlib's pyplot, and seaborn packages as aliases pd,np,plt, and sns respectively
import pandas as pd
import numpy as np
from matplotlib import pyplot as plt
import seaborn as sns

#enables visualizations in line with notebook
%matplotlib inline

<a id='wrangling'></a>
## Data Wrangling

<p>I will Load the data from tbdb-movies.csv file into a pandas Dataframe and perform operations to inspect data types and look for instances of missing or possibly errant data.</p>

<p>I will also inspect the data to know the columns necessary for my analysis.</p>

In [2]:
#loads data from tmdb-movies.csv file and store in df variable
df = pd.read_csv('tmdb-movies.csv')

In [3]:
#Checks the number of entries(rows) and columns
df.shape

(10866, 21)

The table contains 21 columns, and 10,866 rows.

In [4]:
#Checks in each column, the number of columns, column names, column data types, and the number of non-null cells.
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10866 entries, 0 to 10865
Data columns (total 21 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   id                    10866 non-null  int64  
 1   imdb_id               10856 non-null  object 
 2   popularity            10866 non-null  float64
 3   budget                10866 non-null  int64  
 4   revenue               10866 non-null  int64  
 5   original_title        10866 non-null  object 
 6   cast                  10790 non-null  object 
 7   homepage              2936 non-null   object 
 8   director              10822 non-null  object 
 9   tagline               8042 non-null   object 
 10  keywords              9373 non-null   object 
 11  overview              10862 non-null  object 
 12  runtime               10866 non-null  int64  
 13  genres                10843 non-null  object 
 14  production_companies  9836 non-null   object 
 15  release_date       

<p>Columns; imdb_id, cast, homepage, director, tagline, keywords, overview, genres, and production_companies have missing values.</p>
<p>Column; release_date is of object(string) data type instead of datetime data type.</p>


### Data Cleaning

<p>Next, I clean up the data for analysis by eliminating or modifying any data that is erroneous, incomplete, irrelevant, redundant, or improperly formatted. When it comes to data analysis, this data is usually not necessary or beneficial because it can slow down the process or produce inaccurate results.</p>

In [5]:
#removes columns that are non-essential to the analysis
#including imdb_id, tagline, keywords, overview, budget_adj, and revenue_adj
df.drop(['imdb_id', 'homepage', 'tagline', 'keywords', 'overview', 'budget_adj', 'revenue_adj'], axis = 1, inplace = True)

<p>Next, I create a column <i>profit</i> which is necessary for my analysis by subtracting <i>budget</i> from <i>revenue</i>.</p>

In [6]:
#adds a new column 'profit' to the table
df['profit'] = df['revenue'] - df['budget']

<p>Next, I confirm if the changes made have taken effect.</p>

In [7]:
#Checks in each column, the number of columns, column names, column data types, and the number of non-null cells.
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10866 entries, 0 to 10865
Data columns (total 15 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   id                    10866 non-null  int64  
 1   popularity            10866 non-null  float64
 2   budget                10866 non-null  int64  
 3   revenue               10866 non-null  int64  
 4   original_title        10866 non-null  object 
 5   cast                  10790 non-null  object 
 6   director              10822 non-null  object 
 7   runtime               10866 non-null  int64  
 8   genres                10843 non-null  object 
 9   production_companies  9836 non-null   object 
 10  release_date          10866 non-null  object 
 11  vote_count            10866 non-null  int64  
 12  vote_average          10866 non-null  float64
 13  release_year          10866 non-null  int64  
 14  profit                10866 non-null  int64  
dtypes: float64(2), int6

<p>Now, I have total of 15 columns, and all the non-essential columns have been remove.</p>
<p>Next, I check for duplicate entries.</p>

In [8]:
#Checks and returns the total number of duplicate entries.
sum(df.duplicated())

1

<p>The data has just 1 duplicate entry, which I am going to eliminate in the next step.</p>

In [9]:
#drops the row with a duplicate entry
df.drop_duplicates(inplace = True)

<p>Now I confirm if the duplicate row was dropped.</p>

In [10]:
#Checks and returns the total number of duplicate entries.
sum(df.duplicated())

0

<p>Now, the data has no duplicate row.</p>
<p>In the next steps, I will check for values entered as Zero(0), and missing values, then eliminate them.</p>

In [11]:
#checks and returns booleans, indicating whether or not a column has 0 as a value.
df.eq(0).any()

id                      False
popularity              False
budget                   True
revenue                  True
original_title          False
cast                    False
director                False
runtime                  True
genres                  False
production_companies    False
release_date            False
vote_count              False
vote_average            False
release_year            False
profit                   True
dtype: bool

<p>The output says columns; budget, revenue, runtime, and profit have 0 entered as values.</p>
<p>In the next step, I will replace all the zeros in the above mentioned columns with numpy's NaN (not a number).</p>

In [12]:
#assigns a list of columns; budget, revenue, runtime, and profit to a variable; zero_data_cols
zero_data_cols = ['budget', 'revenue', 'runtime', 'profit']

#replaces 0 with numpy's NaN in the list; Zero_data_cols
df[zero_data_cols] = df[zero_data_cols].replace(0, np.NaN)

<p>Next, I check for entries with no value in the table.</p>

In [13]:
#returns the total number of null (no value) values in each columns.
df.isnull().sum()

id                         0
popularity                 0
budget                  5696
revenue                 6016
original_title             0
cast                      76
director                  44
runtime                   31
genres                    23
production_companies    1030
release_date               0
vote_count                 0
vote_average               0
release_year               0
profit                  4705
dtype: int64

<p>The output shows that there are null values in columns; budget, revenue, cast, director, runtime, genres, production_companies, and profit.</p>
<p>In the next step, I will eliminate the rows with null values.</p>

In [14]:
#drops all rows with null values
df.dropna(inplace = True)

<p>Next, I confirm there are no null values in the table.</p>

In [15]:
#returns the total number of null (no value) values in each columns.
df.isnull().sum()

id                      0
popularity              0
budget                  0
revenue                 0
original_title          0
cast                    0
director                0
runtime                 0
genres                  0
production_companies    0
release_date            0
vote_count              0
vote_average            0
release_year            0
profit                  0
dtype: int64

<p>The output shows there are no null values in any column.</p>
<p>Next, I check the details of my data again.</p>

In [16]:
#Checks in each column, the number of columns, column names, column data types, and the number of non-null cells.
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 3802 entries, 0 to 10848
Data columns (total 15 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   id                    3802 non-null   int64  
 1   popularity            3802 non-null   float64
 2   budget                3802 non-null   float64
 3   revenue               3802 non-null   float64
 4   original_title        3802 non-null   object 
 5   cast                  3802 non-null   object 
 6   director              3802 non-null   object 
 7   runtime               3802 non-null   float64
 8   genres                3802 non-null   object 
 9   production_companies  3802 non-null   object 
 10  release_date          3802 non-null   object 
 11  vote_count            3802 non-null   int64  
 12  vote_average          3802 non-null   float64
 13  release_year          3802 non-null   int64  
 14  profit                3802 non-null   float64
dtypes: float64(6), int64

<p>The output shows, there are now 15 columns and 3802 rows in the table, and no null values.</p>
<p>However, column; release_date still shows as object data type instead of datetime data type.</p>
<p>In the next step, I will change the data type of the release_date column to datetime.</p>

In [17]:
df['release_date'] = pd.to_datetime(df['release_date'])

<p>To confirm the effect of the change, I check the details of the data again.</p>

In [18]:
#Checks in each column, the number of columns, column names, column data types, and the number of non-null cells.
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 3802 entries, 0 to 10848
Data columns (total 15 columns):
 #   Column                Non-Null Count  Dtype         
---  ------                --------------  -----         
 0   id                    3802 non-null   int64         
 1   popularity            3802 non-null   float64       
 2   budget                3802 non-null   float64       
 3   revenue               3802 non-null   float64       
 4   original_title        3802 non-null   object        
 5   cast                  3802 non-null   object        
 6   director              3802 non-null   object        
 7   runtime               3802 non-null   float64       
 8   genres                3802 non-null   object        
 9   production_companies  3802 non-null   object        
 10  release_date          3802 non-null   datetime64[ns]
 11  vote_count            3802 non-null   int64         
 12  vote_average          3802 non-null   float64       
 13  release_year     